src/main/java/com/notably/model/block/BlockImpl.java

```diff
@@ -0,0 +1,121 @@
+package com.notably.model.block;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Represents a Block in the Notably data structure.
+ */
+public class BlockImpl implements Block {
+    private List<Block> children;
+    private Title title;
+    private Body body;
+    private Block parent;
+
+    /**
+     * Initializes a default block.
+     * Used when creating a root block and creating a block without the optional body argument.
+     */
+    public BlockImpl(Block parent, Title title) {
+        this.title = title;
+        this.parent = parent;
+    }
+
+    /**
+     * Initializes a block with body content.
+     * Used when creating a block with additional body argument.
+     */
+    public BlockImpl(Block parent, Title title, Body body) {
+        this.title = title;
+        this.parent = parent;
+        this.body = body;
+    }
+
+    @Override
+    public Title getTitle() {
+        return this.title;
+    }
+
+    @Override
+    public Body getBody() {
+        return this.body;
+    }
+
+    @Override
+    public Optional<Block> getParent() {
+        return Optional.ofNullable(parent);
+    }
+
+    @Override
+    public List<Block> getChildren() {
+        return this.children;
+    }
+
+    @Override
+    public void setChildren(List<Block> newChildren) {
+        this.children = newChildren;
+    }
+
+    @Override
+    public Optional<Block> getChild(Title title) {
+        Optional<Block> child = this.children.stream()
+            .filter(block -> title.equals(block.getTitle()))
+            .findAny();
+        return child;
+    }
+
+    @Override
+    public void setChild(Title title, Block newBlock) {
+        Optional<Block> child = this.children.stream()
+            .filter(block -> title.equals(block.getTitle()))
+            .findAny();
+        if (child.isPresent()) {
+            int childIndex = this.children.indexOf(child.get());
+            this.children.set(childIndex, newBlock);
+        }
+    }
+
+    @Override
+    public void addChild(Block block) {
+        if (this.children == null) {
```

Comment:
Hmmm, actually, I feel like it's better to initialize `this.children` to an empty list in the constructor. This prevents the `getChildren()` method from returning `null` (which can lead to potential bugs)

Reply:
One of the reasons why returning a `null` in place of a `List` might lead to bugs is because usually, people expect an empty list rather than a `null`!

Reply:
Got it! I'll make changes